File: Python/03-LongestSubstring.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        
        # length of longest substring
        max_len = 0
        
        # index of current substring whose length is counted
        substring_start = 0
        
        # dictionary containing char:index pairs
        used = {}
        
        for i, char in enumerate(s):
            
            # if a repeated character occurs, new substring starts one
            # index after the last occurence
            if char in used and substring_start <= used[char]:
                substring_start = used[char] + 1
                
            # only updated maximum length if the new substring is longer
            else:
                max_len = max(max_len, i-substring_start+1)
                
            used[char] = i

        return max_len

File: Java/03-LongestSubstring.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    public int lengthOfLongestSubstring(String s) {
    	
    	// length of longest substring
    	int max_len = 0;
    	
    	// index of current substring whose length is counted
    	int substring_start = 0;
    	
    	// hash table containing char:index pairs
    	Map<String, Integer> used = new HashMap<>();
    	
    	int n = s.length();
    	for(int i = 0; i < n; i++) {
    		
    		String si = s.substring(i,i+1);
    		
    		// if a repeated character occurs, new substring starts one
            // index after the last occurrence
    		if (used.containsKey(si) && substring_start <= used.get(si)) {
				substring_start = used.get(si) + 1;
			}
    		else {
    			
    			// only updated maximum length if the new substring is longer
    			max_len = Math.max(max_len, i - substring_start + 1);
    		}
    		
    		// fill hash table until solution is found
    		used.put(si, i);
    		
    	}
		return max_len;
        
    }
}